pkg/cephmgr/osd/partition/scheme.go

```diff
@@ -117,45 +136,258 @@ func (s *Scheme) Save(configDir string) error {
+		SizeMB:        UseRemainingSpace,
+		Offset:        offset,
+	}
+
+	return nil
+}
+
+// populates a partition scheme entry for an OSD that will have distributed partitions: its metadata will live on a
+// dedicated metadata device and its block data will live on a dedicated device
+func PopulateDistributedPerfSchemeEntry(entry *PerfSchemeEntry, device string, metadataInfo *MetadataDeviceInfo,
+	bluestoreConfig BluestoreConfig) error {
+
+	diskUUID, walUUID, dbUUID, blockUUID, err := createUUIDs()
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	return &scheme, nil
+	// the block partition will take up the entire given device
+	entry.Partitions[BlockPartitionName] = &PerfSchemePartitionDetails{
+		Device:        device,
+		DiskUUID:      diskUUID.String(),
+		PartitionUUID: blockUUID.String(),
+		SizeMB:        UseRemainingSpace,
+		Offset:        1,
+	}
+
+	// the WAL and DB will be on a separate metadata device
+	offset := int64(1)
+	numMetadataParts := len(metadataInfo.Partitions)
+	if numMetadataParts == 0 {
+		// the metadata device hasn't been used yet, create a disk UUID for it
+		u, err := uuid.NewRandom()
+		if err != nil {
+			return fmt.Errorf("failed to get metadata disk uuid. %+v", err)
+		}
+		metadataInfo.DiskUUID = u.String()
+	} else {
+		lastEntry := metadataInfo.Partitions[numMetadataParts-1]
+		offset = lastEntry.Offset + lastEntry.SizeMB
+	}
+
+	walSize := WalDefaultSizeMB
+	if bluestoreConfig.WalSizeMB > 0 {
+		walSize = bluestoreConfig.WalSizeMB
+	}
+	dbSize := DBDefaultSizeMB
+	if bluestoreConfig.DatabaseSizeMB > 0 {
+		dbSize = bluestoreConfig.DatabaseSizeMB
+	}
+
+	// record information about the WAL partition
+	entry.Partitions[WalPartitionName] = &PerfSchemePartitionDetails{
+		Device:        metadataInfo.Device,
+		DiskUUID:      metadataInfo.DiskUUID,
+		PartitionUUID: walUUID.String(),
+		SizeMB:        walSize,
+		Offset:        offset,
+	}
+	walPartitionInfo := &MetadataDevicePartition{
+		ID:            entry.ID,
+		OsdUUID:       entry.OsdUUID,
+		Name:          WalPartitionName,
+		PartitionUUID: walUUID.String(),
+		SizeMB:        walSize,
+		Offset:        offset,
+	}
+	metadataInfo.Partitions = append(metadataInfo.Partitions, walPartitionInfo)
+	offset += entry.Partitions[WalPartitionName].SizeMB
+
+	// record information about the DB partition
+	entry.Partitions[DatabasePartitionName] = &PerfSchemePartitionDetails{
+		Device:        metadataInfo.Device,
+		DiskUUID:      metadataInfo.DiskUUID,
+		PartitionUUID: dbUUID.String(),
+		SizeMB:        dbSize,
+		Offset:        offset,
+	}
+	dbPartitionInfo := &MetadataDevicePartition{
+		ID:            entry.ID,
+		OsdUUID:       entry.OsdUUID,
+		Name:          DatabasePartitionName,
+		PartitionUUID: dbUUID.String(),
+		SizeMB:        dbSize,
+		Offset:        offset,
+	}
+	metadataInfo.Partitions = append(metadataInfo.Partitions, dbPartitionInfo)
+
+	return nil
+}
+
+func (m *MetadataDeviceInfo) GetPartitionArgs() []string {
```

Comment:
would it be worth splitting the scheme.go into multiple source files with the classes separate?